Add measurement/timer system and sourceFile to Statistics base class

haved opened the pull request:
The Statistics part of #382. Modifies all existing Statistics to provide the sourceFile path.

The mapping from id to name is now in `Statistics.cpp`, but many of the existing statistics used names that were different from the C++ enums, so I tried to make the classes keep their old names.

It might be better to have `GetName()` be an abstract method, and let subclasses provide names that way, instead of a single BijectiveMap. Let me know!

phate commented on jlm/llvm/opt/InvariantValueRedirection.cpp:
```diff
@@ -19,8 +19,8 @@ class InvariantValueRedirection::Statistics final : public util::Statistics
 
-  Statistics()
-      : util::Statistics(Statistics::Id::InvariantValueRedirection)
+  Statistics(util::filepath sourceFile)
```
`explicit` would be nice.

phate commented on jlm/llvm/opt/cne.cpp:
```diff
@@ -16,11 +16,10 @@ namespace jlm::llvm
 
-  cnestat()
-      : Statistics(Statistics::Id::CommonNodeElimination),
+  cnestat(util::filepath sourceFile)
```
`explicit`

phate commented on jlm/llvm/opt/inlining.cpp:
```diff
@@ -16,11 +16,10 @@ namespace jlm::llvm
 
-  ilnstat()
-      : Statistics(Statistics::Id::FunctionInlining),
+  ilnstat(util::filepath sourceFile)
```
`explicit`

phate commented on jlm/util/Statistics.hpp:
```diff
@@ -62,11 +74,123 @@ class Statistics
+  /**
+   * @return a string identifying the type of this Statistics instance
+   */
+  [[nodiscard]] std::string_view
```
`const std::string_view &` as return type?

phate commented on jlm/util/Statistics.cpp:
```diff
@@ -1,15 +1,171 @@
+  for (auto & [mName, measurement] : Measurements_)
+    if (mName == name)
+      return measurement;
+  throw util::error(util::strfmt("Unknown measurement: ", name));
```
I would almost like to replace this with `JLM_ASSERT()`. Nobody catches it anyway, and it is basically a bug to ask for a measurement that is not present.

phate commented on jlm/util/Statistics.cpp:
```diff
@@ -1,15 +1,171 @@
+  for (auto & [mName, timer] : Timers_)
+    if (mName == name)
+      return timer;
+  throw util::error(util::strfmt("Unknown timer: ", name));
```
`JLM_ASSERT()`

phate commented on jlm/util/Statistics.hpp:
```diff
@@ -62,11 +74,123 @@ class Statistics
+  /**
+   * Adds a measurement, identified by \p name, with the given value.
+   * Requires that the measurement doesn't already exist.
+   * Measurements are listed in insertion order.
```
This is not really relevant for the documentation of this method, right?

phate commented on jlm/util/Statistics.hpp:
```diff
@@ -62,11 +74,123 @@ class Statistics
+    Measurements_.emplace_back(std::make_pair(std::move(name), std::move(value)));
+  }
+
+  // Mutable version of @see GetMeasurement
```
Does doxygen recognize `//`?

phate commented on jlm/util/Statistics.hpp:
```diff
@@ -62,11 +74,123 @@ class Statistics
+  util::timer &
+  AddTimer(std::string name);
+
+  // Mutable version of @see GetTimer
```
Does doxygen recognize `//`?

phate commented on jlm/util/Statistics.hpp:
```diff
@@ -62,11 +74,123 @@ class Statistics
 
 private:
   Statistics::Id StatisticsId_;
+  jlm::util::filepath SourceFile_;
```
`util::filepath SourceFile_`

phate commented on tests/jlm/util/TestStatistics.cpp:
```diff
@@ -14,36 +14,58 @@
+  assert(statistics.GetMeasurementValue<uint64_t>("count") == 10);
+  assert(statistics.GetMeasurementValue<double>("weight") == 6.0);
+  assert(statistics.GetMeasurementValue<int64_t>("bankAccount") == -400);
+  assert(statistics.GetMeasurementValue<std::string>("count") == "poor");
```
I think the key should be `"state"` instead of `"count"`

phate reviewed (dismissed):
Some minor issues, but nothing critical. Up to you whether you want to fix them.

I guess the next steps are (?):
1. Convert all statistics to use the new interface
2. Clean up the Serialize()/ToString() methods of the Statistics class

If you want, I can take care of the (mundane) first point such that you can continue working on Andersen.

haved commented:
Thanks for the comments, I fixed almost all of them. I elected to go for `const util::filepath & sourceFile`, and remove a bunch of `std::move`.

I also modified the set of methods in `Statistics.hpp`, since it turns out C++ does overload resolution before checking visibility. Having public-but-const and protected non-const versions of the same method worked surprisingly poorly.

The only comment you had that I didn't act on was returning `string_view` vs. `const string_view &`, as I think the first is more flexible, and doesn't involve any copying.

Regarding converting classes, I was not planning on doing that work now. I want to get Andersen's statistics over to the `AddMeasurement` functionality, so that it becomes easier to add my new worklist implementation side-by-side.

phate reviewed (approved):
LGTM
